File: packages/core/src/hlc.ts
/**
 * Hybrid Logical Clock — PROTOCOL.md §4
 *
 * State: (millis, counter, deviceId).
 * String form: 12-hex millis + "-" + 4-hex counter + "-" + 16-hex deviceId,
 * fixed 34 chars, lexicographic order === logical order.
 */

export const MAX_COUNTER = 0xffff;
export const MAX_MILLIS = 2 ** 48 - 1;
export const MAX_DRIFT_MS = 300_000; // 5 minutes, §4.5

export interface Hlc {
  readonly millis: number;
  readonly counter: number;
  readonly deviceId: string;
}

export class ClockDriftError extends Error {
  readonly code = "CLOCK_DRIFT";
  constructor(remoteMillis: number, physicalNow: number) {
    super(
      `remote HLC is ${remoteMillis - physicalNow}ms ahead of local clock ` +
        `(max allowed ${MAX_DRIFT_MS}ms); check system time`,
    );
  }
}

const DEVICE_ID_RE = /^[0-9a-f]{16}$/;
const HLC_STRING_RE = /^[0-9a-f]{12}-[0-9a-f]{4}-[0-9a-f]{16}$/;

/** Generate a fresh 64-bit deviceId (16 lowercase hex chars). §4.3 */
export function generateDeviceId(
  randomBytes: (n: number) => Uint8Array = defaultRandomBytes,
): string {
  const b = randomBytes(8);
  let s = "";
  for (const byte of b) s += byte.toString(16).padStart(2, "0");
  return s;
}

function defaultRandomBytes(n: number): Uint8Array {
  const b = new Uint8Array(n);
  // globalThis.crypto exists in browsers, Node >= 19, Deno, Bun, workers.
  globalThis.crypto.getRandomValues(b);
  return b;
}

export function isValidDeviceId(s: string): boolean {
  return DEVICE_ID_RE.test(s);
}

/** Canonical string encoding. §4.2 */
export function encodeHlc(h: Hlc): string {
  return (
    h.millis.toString(16).padStart(12, "0") +
    "-" +
    h.counter.toString(16).padStart(4, "0") +
    "-" +
    h.deviceId
  );
}

/** Parse canonical string form; throws on malformed input. */
export function decodeHlc(s: string): Hlc {
  if (!HLC_STRING_RE.test(s)) {
    throw new Error(`malformed HLC string: ${JSON.stringify(s)}`);
  }
  return {
    millis: parseInt(s.slice(0, 12), 16),
    counter: parseInt(s.slice(13, 17), 16),
    deviceId: s.slice(18),
  };
}

/**
 * Total order over HLCs: millis, then counter, then deviceId.
 * Guaranteed to agree with bytewise comparison of encodeHlc output;
 * the property tests verify this equivalence.
 */
export function compareHlc(a: Hlc, b: Hlc): number {
  if (a.millis !== b.millis) return a.millis < b.millis ? -1 : 1;
  if (a.counter !== b.counter) return a.counter < b.counter ? -1 : 1;
  if (a.deviceId !== b.deviceId) return a.deviceId < b.deviceId ? -1 : 1;
  return 0;
}

/**
 * The clock itself. Owns mutable (millis, counter) state for one device in
 * one space. Inject `physicalClock` for tests; defaults to Date.now.
 *
 * Persistence: callers should persist state() after issuing ops and restore
 * with the constructor, so a restart never reissues a timestamp.
 */
export class HybridLogicalClock {
  private millis: number;
  private counter: number;
  readonly deviceId: string;
  private readonly physicalClock: () => number;

  constructor(opts: {
    deviceId: string;
    physicalClock?: () => number;
    /** restore persisted state; defaults to zero */
    millis?: number;
    counter?: number;
  }) {
    if (!isValidDeviceId(opts.deviceId)) {
      throw new Error(`invalid deviceId: ${JSON.stringify(opts.deviceId)}`);
    }
    this.deviceId = opts.deviceId;
    this.physicalClock = opts.physicalClock ?? Date.now;
    this.millis = opts.millis ?? 0;
    this.counter = opts.counter ?? 0;
    this.checkBounds();
  }

  /** Current state, for persistence. Does not advance the clock. */
  state(): Hlc {
    return { millis: this.millis, counter: this.counter, deviceId: this.deviceId };
  }

  /**
   * Issue a timestamp for a new local op. §4.1 "send/local event".
   * Strictly greater than every timestamp this clock has issued or observed.
   */
  now(): Hlc {
    const pt = this.physicalClock();
    if (pt > this.millis) {
      this.millis = pt;
      this.counter = 0;
    } else {
      this.counter += 1;
      if (this.counter > MAX_COUNTER) {
        // overflow rollover, §4.1
        this.millis += 1;
        this.counter = 0;
      }
    }
    this.checkBounds();
    return this.state();
  }

  /**
   * Observe a remote timestamp. §4.1 "receive".
   * Throws ClockDriftError if the remote is > MAX_DRIFT_MS ahead of our
   * physical clock (§4.5). On success the local clock becomes strictly
   * greater than both its previous state and the remote timestamp.
   */
  receive(remote: Hlc): Hlc {
    const pt = this.physicalClock();
    if (remote.millis > pt + MAX_DRIFT_MS) {
      throw new ClockDriftError(remote.millis, pt);
    }
    const m = Math.max(this.millis, remote.millis, pt);
    let c: number;
    if (m === this.millis && m === remote.millis) {
      c = Math.max(this.counter, remote.counter) + 1;
    } else if (m === this.millis) {
      c = this.counter + 1;
    } else if (m === remote.millis) {
      c = remote.counter + 1;
    } else {
      c = 0;
    }
    this.millis = m;
    this.counter = c;
    if (this.counter > MAX_COUNTER) {
      this.millis += 1;
      this.counter = 0;
    }
    this.checkBounds();
    return this.state();
  }

  private checkBounds(): void {
    if (this.millis > MAX_MILLIS) {
      throw new Error("HLC millis exceeded 48-bit range");
    }
  }
}

File: impl/dart/lib/src/hlc.dart
/// Hybrid Logical Clock — PROTOCOL.md section 4.
///
/// The encoding is what everything else leans on: fixed-width lowercase hex
/// so that plain string comparison equals logical comparison, which lets
/// storage engines sort and range-scan HLCs as opaque strings.
library;

const int maxCounter = 0xffff;
const int maxMillis = 0xffffffffffff; // 2^48 - 1
const int maxDriftMs = 300000; // 5 minutes, section 4.5

/// Thrown when a remote timestamp is further ahead than section 4.5 allows.
///
/// The whole batch fails: partial application would leave the frontier
/// claiming ops that were dropped.
class ClockDriftError implements Exception {
  ClockDriftError(this.remoteMillis, this.physicalMillis);

  final int remoteMillis;
  final int physicalMillis;

  @override
  String toString() =>
      'ClockDriftError: remote clock is ${remoteMillis - physicalMillis}ms '
      'ahead (max allowed ${maxDriftMs}ms); check system time';
}

final RegExp _deviceIdPattern = RegExp(r'^[0-9a-f]{32}$');
final RegExp _hlcPattern = RegExp(r'^([0-9a-f]{12})-([0-9a-f]{4})-([0-9a-f]{32})$');

bool isValidDeviceId(String id) => _deviceIdPattern.hasMatch(id);

/// An immutable timestamp: (millis, counter, deviceId).
class Hlc implements Comparable<Hlc> {
  const Hlc(this.millis, this.counter, this.deviceId);

  final int millis;
  final int counter;
  final String deviceId;

  /// Section 4.2. Fixed width 50: 12 + 1 + 4 + 1 + 32.
  String encode() {
    final m = millis.toRadixString(16).padLeft(12, '0');
    final c = counter.toRadixString(16).padLeft(4, '0');
    return '$m-$c-$deviceId';
  }

  static Hlc decode(String s) {
    final match = _hlcPattern.firstMatch(s);
    if (match == null) {
      throw FormatException('malformed HLC string', s);
    }
    return Hlc(
      int.parse(match.group(1)!, radix: 16),
      int.parse(match.group(2)!, radix: 16),
      match.group(3)!,
    );
  }

  /// Logical ordering: millis, then counter, then deviceId.
  ///
  /// Guaranteed to agree with `encode().compareTo(other.encode())`, because
  /// every field is zero-padded to a fixed width and Dart compares strings by
  /// UTF-16 code unit.
  @override
  int compareTo(Hlc other) {
    if (millis != other.millis) return millis < other.millis ? -1 : 1;
    if (counter != other.counter) return counter < other.counter ? -1 : 1;
    return deviceId.compareTo(other.deviceId);
  }

  @override
  bool operator ==(Object other) =>
      other is Hlc &&
      other.millis == millis &&
      other.counter == counter &&
      other.deviceId == deviceId;

  @override
  int get hashCode => Object.hash(millis, counter, deviceId);

  @override
  String toString() => encode();
}

/// A device's clock for one space. Section 4.1.
class HybridLogicalClock {
  HybridLogicalClock({
    required this.deviceId,
    int Function()? physicalClock,
    int millis = 0,
    int counter = 0,
  })  : _physicalClock = physicalClock ?? _wallClock,
        _millis = millis,
        _counter = counter {
    if (!isValidDeviceId(deviceId)) {
      throw ArgumentError.value(deviceId, 'deviceId', 'must be 16 lowercase hex chars');
    }
    _checkBounds();
  }

  static int _wallClock() => DateTime.now().millisecondsSinceEpoch;

  final String deviceId;
  final int Function() _physicalClock;
  int _millis;
  int _counter;

  /// Current state, for persistence. Does not advance the clock.
  Hlc get state => Hlc(_millis, _counter, deviceId);

  /// Issue a timestamp for a new local op. Section 4.1 "send/local event".
  Hlc now() {
    final pt = _physicalClock();
    if (pt > _millis) {
      _millis = pt;
      _counter = 0;
    } else {
      _counter += 1;
      if (_counter > maxCounter) {
        _millis += 1;
        _counter = 0;
      }
    }
    _checkBounds();
    return state;
  }

  /// Observe a remote timestamp. Section 4.1 "receive".
  ///
  /// Afterwards the local clock is strictly greater than both its previous
  /// value and [remote], which is what preserves causality.
  Hlc receive(Hlc remote) {
    final pt = _physicalClock();
    if (remote.millis > pt + maxDriftMs) {
      throw ClockDriftError(remote.millis, pt);
    }

    final m = [_millis, remote.millis, pt].reduce((a, b) => a > b ? a : b);
    final int c;
    if (m == _millis && m == remote.millis) {
      c = (_counter > remote.counter ? _counter : remote.counter) + 1;
    } else if (m == _millis) {
      c = _counter + 1;
    } else if (m == remote.millis) {
      c = remote.counter + 1;
    } else {
      c = 0;
    }

    _millis = m;
    _counter = c;
    if (_counter > maxCounter) {
      _millis += 1;
      _counter = 0;
    }
    _checkBounds();
    return state;
  }

  void _checkBounds() {
    if (_millis > maxMillis) {
      throw StateError('HLC millis exceeded 48-bit range');
    }
  }
}
